_posts/2022-09-XX-rate-limiting-web-applications.markdown

```diff
@@ -0,0 +1,465 @@
+
+### Why rate limiting?
+
+Let's say you are building a web API that lets you store todo items.
+Nice and simple: a `GET /api/todos` that returns a list of todo items, and a `POST /api/todos` that let you create and update a specific todo item.
+What could possibly go wrong with using these two endpoints?
+
+Off the top of my head:
+* The mobile app another team is building accidentally causes an infinite loop that keeps calling `POST`, and tries to create a new todo item 10.000 times over the course of a few seconds before it crashes. That's a lot of todo items in the database that should not be there.
+* Someone is brute-forcing the `GET` method, trying to get todo items for all your users. You have security in place, so they will never get in without valid credentials, but your database has to run a query to check credentials 10 times per second. That's rough on this small 0.5 vCPU database instance that seemed good on paper.
+* An aggressive search engine spider accidentally adding 20.000 items into a shopping cart that is stored in memory.
+
+There are probably more things that could go wrong, but you get the picture.
+You, your team, or external factors may behave in ways you did not expect.
+
+That profile picture upload that usually gets small images uploaded? Guaranteed someone will try to upload a 500MB picture of the universe at some point.
+
+When you build an application, there's a very real chance that you don't know how it will be used, and what potential abuse may look like.
+You are sharing CPU, memory and database usage among your users, One bad actor, whether intentional or accidental, can break or make your application slow, spoiling the experience for other users.
+
+### What is rate limiting?
+
+Rate limiting, or request throttling, is an approach to reduce the fall-out of unexpected or unwanted traffic patterns to your application.
+
+Typically, web applications implement rate limiting by setting an allowance on the number of requests for a given timeframe.
+If you are a streaming service, you may want to limit the outgoing bandwidth per user over a given time. Up to you!
+
+The ultimate goal of imposing rate limits is to reduce or even eliminate traffic and usage of your application that is potentially damaging.
+Regardless of the traffic being accidental or malicious.
+
+### What should I rate limit?
+
+I will give you a quote that you can use in other places:
+
+> *Rate limit everything.*<br/>
+> *– Maarten Balliauw*
+
+With *everything*, I mean every endpoint that uses resources that could slow down or break your application when exhausted or stressed.
+
+Typically, you'll want to rate limit endpoints that make use of the CPU, memory, disk I/O, the database, external APIs, and the likes.
+
+Huh. That does mean everything! You'll want to prevent resource exhaustion, and make usage of shared resources more fair to all your users.
```

Comment:
```suggestion
Huh. That does mean everything, even your internal (health) endpoints! You'll want to prevent resource exhaustion, and make usage of shared resources more fair to all your users.
```